ql/src/java/org/apache/hadoop/hive/ql/queryhistory/QueryHistoryService.java

```diff
@@ -0,0 +1,266 @@
+    } else {
+      LOG.info("Query History max queue memory byte size is {}: records will be persisted when the queue consumes " +
+              "more than {} bytes in the memory",
+          maxQueueSizeInMemoryBytes, maxQueueSizeInMemoryBytes);
+    }
+  }
+
+  public static QueryHistoryPersistor createPersistor(HiveConf conf) {
+    initializeSessionForTableCreation(conf);
+
+    try {
+      return (QueryHistoryPersistor) ReflectionUtil.newInstance(
+          conf.getClassByName(conf.get(HiveConf.ConfVars.HIVE_QUERY_HISTORY_SERVICE_PERSISTOR_CLASS.varname)), conf);
+    } catch (ClassNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private static void initializeSessionForTableCreation(HiveConf conf) {
+    HiveConf.setVar(conf, HiveConf.ConfVars.HIVE_QUERY_ID, QueryHistoryPersistor.QUERY_ID_FOR_TABLE_CREATION);
+    SessionState ss = SessionState.get(); // if there is a SessionState in the called thread, we can use that
+    if (ss == null){
+      ss = SessionState.start(conf);
+    }
+    ss.addQueryState(QueryHistoryPersistor.QUERY_ID_FOR_TABLE_CREATION,
+        new QueryState.Builder().withHiveConf(conf).build());
+  }
+
+  // keeping a global, static instance is usually considered an antipattern, even in case of using the Singleton
+  // design, this time the constraint was the ability to hook into the query processing (Driver) from somewhere the
+  // outer service (HiveServer2, QTestUtil, etc.), which otherwise would have needed passing a service instance
+  // through many layers, making the code and method signatures more complicated
+  public static void handle(DriverContext driverContext) {
+    if (INSTANCE == null){
+      // we can only hit this codepath if hive.query.history.service.enabled=true but the INSTANCE is not initialized,
+      // which is unlikely to happen under a production HiveServer2, as it initializes the service properly
+      LOG.warn("QueryHistoryService is called but not initialized, this is only fine under testing circumstances.");
+      return;
+    }
+    INSTANCE.handleRecord(driverContext);
+  }
+
+  @VisibleForTesting
+  protected QueryHistoryService() {
+  }
+
+  private void persistRecordsIfNeeded() {
+    if (needToPersist()) {
+      LOG.debug("Submitting a job to persist {} history records", queryHistoryQueue.size());
+      persistExecutor.submit(() -> {
+        queryHistoryPersistor.persist(queryHistoryQueue);
+      });
+      if (maxBatchSize == 0) {
+        waitForAllRecordsToBePersisted();
+      }
+    } else {
+      LOG.debug("Not persisting history records yet, current queue size: {}, maxBatchSize: {}",
+          queryHistoryQueue.size(),
+          maxBatchSize);
+    }
+  }
+
+  private void waitForAllRecordsToBePersisted() {
+    while (!queryHistoryQueue.isEmpty()){
+      try {
+        LOG.info("Waiting for {} records to persist synchronously", queryHistoryQueue.size());
+        Thread.sleep(500);
```

Comment:
that is a red flag and anti-pattern, please use completablefuture/notify etc

Reply:
okay, I usually use CountDownLatch for this purpose, let me try it in this scenario